File: lua/sum-of-multiples/sum-of-multiples.lua
return function(numbers)
  local multiples = {}

  local function reduce(list)
    local sum = 0

    for n,_ in pairs(list) do
      sum = sum + n
    end

    return sum
  end

  local function to(limit)
    for i=1,#numbers do
      local num = numbers[i]
      local current = num

      while current < limit do
        if not multiples[current] then
          multiples[current] = true
        end
        current = current + num
      end
    end

    return reduce(multiples)
  end

  return { to = to }
end

File: clojure/sum-of-multiples/src/sum_of_multiples.clj
(ns sum-of-multiples)

(defn multiples-of-n-to-upper-limit
  "Calculate multiples of n up to but not including limit"
  [n limit]
  (take-while #(> limit %) (iterate (partial + n) n)))

(defn sum-of-multiples
  "Calculate sum of unique multiples"
  [list1 limit]
  (->> list1
       (mapcat #(multiples-of-n-to-upper-limit % limit))
       (distinct)
       (reduce +)))
